datumaro/components/project.py

```diff
@@ -2,486 +2,2389 @@
+    paths = {osp.join(repo_root, osp.normpath(p)): _make_ignored_path(p)
+        for p in paths}
+
+    openmode = 'r+'
+    if not osp.isfile(filepath):
+        openmode = 'w+' # r+ cannot create, w+ truncates
+    with open(filepath, openmode) as f:
+        lines = []
+        if mode in {IgnoreMode.append, IgnoreMode.remove}:
+            for line in f:
+                lines.append(line.strip())
+            f.seek(0)
+
+        new_lines = []
+        for line in lines:
+            if not line or line.startswith('#'):
+                new_lines.append(line)
+                continue
+
+            line_path = osp.join(repo_root,
+                osp.normpath(line.split('#', maxsplit=1)[0]).lstrip('/'))
+
+            if mode == IgnoreMode.append:
+                if line_path in paths:
+                    paths.pop(line_path)
+                new_lines.append(line)
+            elif mode == IgnoreMode.remove:
+                if line_path in paths:
+                    paths.pop(line_path)
+                else:
+                    new_lines.append(line)
+
+        if mode in { IgnoreMode.rewrite, IgnoreMode.append }:
+            new_lines.extend(paths.values())
+
+        if not new_lines or new_lines[0] != header:
+            f.write(header + '\n')
+        f.writelines('\n'.join(new_lines))
+        f.truncate()
+
+class CrudProxy:
+    @property
+    def _data(self):
+        raise NotImplementedError()
+
+    def __len__(self):
+        return len(self._data)
+
+    def __getitem__(self, name):
+        return self._data[name]
+
+    def get(self, name, default=None):
+        return self._data.get(name, default)
+
+    def __iter__(self):
+        return iter(self._data.keys())
+
+    def items(self):
+        return iter(self._data.items())
+
+    def __contains__(self, name):
+        return name in self._data
+
+class _DataSourceBase(CrudProxy):
+    def __init__(self, project, config_field):
         self._project = project
-        self._env = project.env
-        config = self.config
-        env = self.env
+        self._field = config_field
+
+    @CrudProxy._data.getter
+    def _data(self):
+        return self._project.config[self._field]
+
+    def add(self, name, value):
+        if name in self:
+            raise SourceExistsError(name)
 
-        sources = {}
-        for s_name, source in config.sources.items():
-            s_format = source.format
+        return self._data.set(name, value)
 
-            url = source.url
-            if not source.url:
-                url = osp.join(config.project_dir, config.sources_dir, s_name)
+    def remove(self, name):
+        self._data.remove(name)
+
+class ProjectSources(_DataSourceBase):
+    def __init__(self, project):
+        super().__init__(project, 'sources')
+
+    def __getitem__(self, name):
+        try:
+            return super().__getitem__(name)
+        except KeyError:
+            raise KeyError("Unknown source '%s'" % name)
```

Comment:
```suggestion
        except KeyError as e:
            raise KeyError("Unknown source '%s'" % name) from e
```

Reply:
How is it helpful here?

Reply:
> How is it helpful here?

It sets the original exception as the cause for the new exception, so that if the exception is printed, the traceback shows the causal relationship. Otherwise, the traceback says that the new exception occurred "during handling of the above exception", which doesn't suggest that they are related.